core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala

```diff
@@ -132,10 +133,15 @@ private[spark] object BlockManagerId {
     getCachedBlockManagerId(obj)
   }
 
-  val blockManagerIdCache = new ConcurrentHashMap[BlockManagerId, BlockManagerId]()
+  val blockManagerIdCache = new TimeStampedHashMap[BlockManagerId, BlockManagerId](true)
 
-  def getCachedBlockManagerId(id: BlockManagerId): BlockManagerId = {
+  def getCachedBlockManagerId(id: BlockManagerId, clearOldValues: Boolean = false): BlockManagerId =
+  {
     blockManagerIdCache.putIfAbsent(id, id)
-    blockManagerIdCache.get(id)
+    val blockManagerId = blockManagerIdCache.get(id)
+    if (clearOldValues) {
+      blockManagerIdCache.clearOldValues(System.currentTimeMillis - Utils.timeStringAsMs("10d"))
```

Comment:
10 days? I don't think *time* can be a judging criteria to decide whether we should remove a cached id or not, even if you set the time threshold far less/greater than '10d'. Think about a extreamly case that a block could be frequently got all the time during the app‘s running. So, it would be certainly removed from cache due to the time threshold, and recached next time we get it, and repeatedly.

Reply:
@Ngone51 Thanks.i also though about remove when we delete a block.
In this case, it is history replaying which will trigger this problem,and we do not delete any block actually.
Maybe use `weakreference` better as @jiangxb1987 mentioned?WDYT?
Thanks again!

Reply:
Can we simply use `com.google.common.cache.Cache`? which has a size limitation and we don't need to worry about OOM.